# **Task 4.9: Part 1**

## **Content**
> #### 01. Importing Libraries
> #### 02. Importing Data
> #### 03. Exploring New Dataset
> #### 04. Data Wrangling
>> ##### 04.1. Data Type Corrections
>> ##### 04.2. Column Name Changes
> #### 05. Data Consistency Checks
>> ##### 05.1. Missing Values
>> ##### 05.2. Addressing Missing Values
>> ##### 05.3. Duplicates
> #### 06. Exporting Customer Dataframe
> #### 07. Combining Datasets
>> ##### 07.1. Importing Prepared Data
>> ##### 07.2. Exploring Datasets
>> ##### 07.3. Merging Datasets
> #### 08. Exporting Combined Dataset

## 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [2]:
# Creating a path to main project folder
path = r'C:\Users\jboer\OneDrive\Documents\Career Foundry\Instacart Basket Analysis'

In [3]:
# Importing "customers" data
cus = pd.read_csv(os.path.join(path,'02 Data','Original Data','customers.csv'))

## 03. Exploring New Dataset

In [4]:
cus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
cus.shape

(206209, 10)

In [6]:
cus.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [13]:
cus.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [24]:
# Checking frequency of "date_joined" column
cus['date_joined'].value_counts(dropna = False)

date_joined
2018-09-17    213
2018-02-10    212
2019-04-01    211
2019-09-21    211
2017-12-19    210
             ... 
2018-09-01    141
2018-01-22    140
2017-11-24    139
2019-07-18    138
2018-08-06    128
Name: count, Length: 1187, dtype: int64

## 04. Data Wrangling

### 04.1. Data Type Corrections

In [20]:
# Changing "date_joined" column from mixed data type to datetime type
cus['date_joined'] = pd.to_datetime(cus['date_joined'])

In [34]:
# Checking dataframe after changing data type
cus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48  2017-01-01   
1    33890   Patricia      Hart  Female  New Mexico   36  2017-01-01   
2    65803    Kenneth    Farley    Male       Idaho   35  2017-01-01   
3   125935   Michelle     Hicks  Female        Iowa   40  2017-01-01   
4   130797        Ann   Gilmore  Female    Maryland   26  2017-01-01   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [26]:
# Checking changed data type
cus.dtypes

user_id                  int64
First Name              object
Surnam                  object
Gender                  object
STATE                   object
Age                      int64
date_joined     datetime64[ns]
n_dependants             int64
fam_status              object
income                   int64
dtype: object

In [27]:
# Checking and comparing frequency of "date_joined" column
cus['date_joined'].value_counts(dropna = False)

date_joined
2018-09-17    213
2018-02-10    212
2019-04-01    211
2019-09-21    211
2017-12-19    210
             ... 
2018-09-01    141
2018-01-22    140
2017-11-24    139
2019-07-18    138
2018-08-06    128
Name: count, Length: 1187, dtype: int64

##### **Note** - after changing the data type of "date_joined" to datetime type, the date format was changed from MM/DD/YYYY to the default YYYY-MM-DD 

### 04.2. Column Name Changes

In [29]:
# Changing name of "First Name" column to "first_name" to match format of other column names
cus.rename(columns={'First Name' : 'first_name'}, inplace = True)

In [30]:
# Changing name of "Surnam" column to "last_name" to make it a more logical name and match format of other column names
cus.rename(columns={'Surnam' : 'last_name'}, inplace = True)

In [31]:
# Changing name of "Gender" column to "gender" to match format of other column names
cus.rename(columns={'Gender' : 'gender'}, inplace = True)

In [32]:
# Changing name of "STATE" column to "state" to match format of other column names
cus.rename(columns={'STATE' : 'state'}, inplace = True)

In [33]:
# Changing name of "Age" column to "age" to match format of other column names
cus.rename(columns={'Age' : 'age'}, inplace = True)

In [35]:
# Checking success of column name changes
cus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48  2017-01-01   
1    33890   Patricia      Hart  Female  New Mexico   36  2017-01-01   
2    65803    Kenneth    Farley    Male       Idaho   35  2017-01-01   
3   125935   Michelle     Hicks  Female        Iowa   40  2017-01-01   
4   130797        Ann   Gilmore  Female    Maryland   26  2017-01-01   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

##### **Note** - No columns will be dropped as all could be useful in my analysis

## 05. Data Consistency Checks

### 05.1. Missing Values

In [36]:
# Checking for missing values
cus.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [37]:
# Creating new dataframe to see missing values
cus_nan = cus[cus['first_name'].isnull() == True]

In [38]:
# Viewing new dataframe "cus_nan"
cus_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26  2017-01-01   
73        13738        NaN     Frost  Female     Louisiana   39  2017-01-01   
82        89996        NaN    Dawson  Female        Oregon   52  2017-01-01   
99        96166        NaN   Oconnor    Male      Oklahoma   51  2017-01-01   
105       29778        NaN    Dawson  Female          Utah   63  2017-01-01   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28  2020-03-31   
206044   200799        NaN  Copeland  Female        Hawaii   52  2020-04-01   
206090   167394        NaN     Frost  Female        Hawaii   61  2020-04-01   
206162   187532        NaN     Floyd  Female    California   39  2020-04-01   
206171   116898        NaN   Delgado  Female      Colorado   23  2020-04-01   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

##### There are 11,259 missing values in the "first_name" column. I do not want to get rid of these columns because these rows still include info in all the other columns that could be useful to my analysis. But, I still want to address the missing values. Since we have information on gender and family status, I want to impute values for any that are missing in the "first_name" column in the following way: "Mr." for any row where gender is "Male", "Mrs." for any row where gender is "Female" and family status is "married", and "Ms." for any row where gender is "Female" and family status is "single". 

### 05.2. Addressing Missing Values

##### I will create a new dataframe (cus_1) and use the loc() function to set conditions to impute values, in the way I mentioned in the markdown above, for the missing values in the "first_name" column.

In [50]:
# Creating a new dataframe by copying original dataframe 
cus_1 = cus.copy()

In [51]:
# Setting condition to change null values to "Mr." for rows where gender is "Male"
cus_1.loc[(cus_1['first_name'].isnull()) & (cus_1['gender'] == 'Male'), 'first_name'] = 'Mr.'

In [52]:
# Setting condition to change null values to "Mrs." for rows where gender is "Female" and fam_status is "married"
cus_1.loc[(cus_1['first_name'].isnull()) & (cus_1['gender'] == 'Female') & (cus_1['fam_status'] == 'married'), 'first_name'] = 'Mrs.'

In [53]:
# Setting condition to change null values to "Ms." for rows where gender is "Female" and fam_status is "single"
cus_1.loc[(cus_1['first_name'].isnull()) & (cus_1['gender'] == 'Female') & (cus_1['fam_status'] == 'single'), 'first_name'] = 'Ms.'

In [54]:
# Checking success of replacing missing values by locating one example of each
print(cus_1.loc[[53, 73, 82]])

    user_id first_name last_name  gender      state  age date_joined  \
53    76659        Mr.   Gilbert    Male   Colorado   26  2017-01-01   
73    13738        Ms.     Frost  Female  Louisiana   39  2017-01-01   
82    89996       Mrs.    Dawson  Female     Oregon   52  2017-01-01   

    n_dependants fam_status  income  
53             2    married   41709  
73             0     single   82518  
82             3    married  117099  


### 05.3. Duplicates

In [55]:
# Checking "cus_1" dataframe for full duplicates
cus_dups = cus_1[cus_1.duplicated()]

In [56]:
cus_dups.shape

(0, 10)

##### There are no full duplicates in my "cus_1" dataframe

In [81]:
# Checking shape of "cus_1" after consistency checks
cus_1.shape

(206209, 10)

## 06. Exporting Customer Dataframe

In [57]:
# Exporting wrangled and checked "cus_1" dataframe
cus_1.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_clean.pkl'))

## 07. Combining Datasets

### 07.1. Importing Prepared Data

In [58]:
# Importing my prepared "orders_products" data to combine with "customer" data
ords_prods = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_agg_final.pkl'))

### 07.2. Exploring Datasets

In [59]:
# Checking "ords_prods" dataframe for common key
ords_prods.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  ...        price_label  \
0                                     Soda  ...  Mid-range product   
1  Organic Unsweetened Vanilla Almond Milk  ...  Mid-range product   
2                      Original Beef Jerky  ...  Low-range product   
3               Aged White Cheddar Popcorn  ...  Low-range product   
4         XL Pick-A-Size Paper Towel Rolls  ...  Low-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy  Regularly busy        Average orders        10   
2  Regularly busy  Regularly busy        Average orders        10   
3  Regularly busy  Regularly busy        Average orders        10   
4  Regularly busy  Regularly busy        Average orders        10   

   loyalty_flag avg_spent spending_flag  median_order_frequency  \
0  New customer  6.367797   Low spender                    20.0   
1  New customer  6.367797   Low spender                    20.0   
2  New customer  6.367797   Low spender                    20.0   
3  New customer  6.367797   Low spender                    20.0   
4  New customer  6.367797   Low spender                    20.0   

  order_frequency_flag  
0     Regular customer  
1     Regular customer  
2     Regular customer  
3     Regular customer  
4     Regular customer  

[5 rows x 24 columns]

In [60]:
# Checking "cus_1" for common key
cus_1.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48  2017-01-01   
1    33890   Patricia      Hart  Female  New Mexico   36  2017-01-01   
2    65803    Kenneth    Farley    Male       Idaho   35  2017-01-01   
3   125935   Michelle     Hicks  Female        Iowa   40  2017-01-01   
4   130797        Ann   Gilmore  Female    Maryland   26  2017-01-01   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [61]:
# Checking shape of "ords_prods"
ords_prods.shape

(32404859, 24)

In [62]:
# Checking shape of "cus_1"
cus_1.shape

(206209, 10)

In [63]:
print(ords_prods.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_label', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_spent', 'spending_flag', 'median_order_frequency',
       'order_frequency_flag'],
      dtype='object')


##### The only common key between the dataframes is "user_id"

### 07.3. Merging Datasets

In [66]:
# Testing inner join merge with key "user_id"
pd.merge(ords_prods, cus_1, on = 'user_id', indicator = True)

ValueError: Cannot use name of an existing column for indicator column

In [71]:
# Dropping "_merge" column in "ords_prods" so it doens't interfere with merging customer data
ords_prods = ords_prods.drop(columns = ['_merge'])

In [72]:
# Checking that "_merge" column was dropped
print(ords_prods.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_spent',
       'spending_flag', 'median_order_frequency', 'order_frequency_flag'],
      dtype='object')


In [73]:
# Re-testing inner join merge with key "user_id"
pd.merge(ords_prods, cus_1, on = 'user_id', indicator = True)

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32404854   2977660   206209            13                   1   
32404855   2977660   206209            13                   1   
32404856   2977660   206209            13                   1   
32404857   2977660   206209            13                   1   
32404858   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     0.0         196   
1                         8                     0.0       14084   
2                         8                     0.0       12427   
3                         8                     0.0       26088   
4                         8                     0.0       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  ...  first_name  last_name  \
0                                            Soda  ...       Linda     Nguyen   
1         Organic Unsweetened Vanilla Almond Milk  ...       Linda     Nguyen   
2                             Original Beef Jerky  ...       Linda     Nguyen   
3                      Aged White Cheddar Popcorn  ...       Linda     Nguyen   
4                XL Pick-A-Size Paper Towel Rolls  ...       Linda     Nguyen   
...                                           ...  ...         ...        ...   
32404854                             Tomato Paste  ...       Diana    Donovan   
32404855          Brownie Crunch High Protein Bar  ...       Diana    Donovan   
32404856    High Protein Bar Chunky Peanut Butter  ...       Diana    Donovan   
32404857      Chocolate Peanut Butter Protein Bar  ...       Diana    Donovan   
32404858      Roasted & Salted Shelled Pistachios  ...       Diana    Donovan   

          gender    state age date_joined n_dependants  fam_status  income  \
0         Female  Alabama  31  2019-02-17            3     married   40423   
1         Female  Alabama  31  2019-02-17            3     married   40423   
2         Female  Alabama  31  2019-02-17            3     married   40423   
3         Female  Alabama  31  2019-02-17            3     married   40423   
4         Female  Alabama  31  2019-02-17            3     married   40423   
...          ...      ...  ..         ...          ...         ...     ...   
32404854  Female     Iowa  74  2019-09-14            3     married  137969   
32404855  Female     Iowa  74  2019-09-14            3     married  137969   
32404856  Female     Iowa  74  2019-09-14            3     married  137969   
32404857  Female     Iowa  74  2019-09-14            3     married  137969   
32404858  Female     Iowa  74  2019-09-14            3     married  137969   

         

##### Test merge successful. Will now merge "ords_prods" and "cus_1" into new dataframe, "ords_prods_cus"

In [74]:
# Merging "cus_1" with "ords_prods" using key "user_id"
ords_prods_cus = ords_prods.merge(cus_1, on = 'user_id', indicator = True)

In [75]:
# Checking merged dataframe "ords_prods_cus"
ords_prods_cus.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   

                              product_name  ...  first_name  last_name  \
0                                     Soda  ...       Linda     Nguyen   
1  Organic Unsweetened Vanilla Almond Milk  ...       Linda     Nguyen   
2                      Original Beef Jerky  ...       Linda     Nguyen   
3               Aged White Cheddar Popcorn  ...       Linda     Nguyen   
4         XL Pick-A-Size Paper Towel Rolls  ...       Linda     Nguyen   

   gender    state age date_joined n_dependants  fam_status income  _merge  
0  Female  Alabama  31  2019-02-17            3     married  40423    both  
1  Female  Alabama  31  2019-02-17            3     married  40423    both  
2  Female  Alabama  31  2019-02-17            3     married  40423    both  
3  Female  Alabama  31  2019-02-17            3     married  40423    both  
4  Female  Alabama  31  2019-02-17            3     married  40423    both  

[5 rows x 33 columns]

In [76]:
# Checking shape of "ords_prods_cus"
ords_prods_cus.shape

(32404859, 33)

In [77]:
# Checking columns in "ords_prods_cus"
print(ords_prods_cus.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_label', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_spent',
       'spending_flag', 'median_order_frequency', 'order_frequency_flag',
       'first_name', 'last_name', 'gender', 'state', 'age', 'date_joined',
       'n_dependants', 'fam_status', 'income', '_merge'],
      dtype='object')


In [78]:
# Confirming results of merge by checking "_merge" frequency
ords_prods_cus['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

## 08. Exporting Combined Dataset

In [80]:
# Exporting final merged "ords_prods_cus" dataframe
ords_prods_cus.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all.pkl'))